Fix/slider input validation

jamesducky opened the pull request:
[Closes [BUG] Slider - Number input validation too restrictive](https://app.asana.com/0/1198919072280397/1202546970461155/f)
[Matches updated UX spec](https://www.figma.com/file/4ZukYZAMHrsCIsrZu9u2a6/Plumage-Components?node-id=405%3A36123)

Changes
- Number Input field now validates on blur, not on change
- Refactors valid input to be called once, and set in state.
- Refactors validation logic to remove variable mutation
- Calls the permitted input range in state once

Issues
- Tests for update validation removed as validation does not happen until blur on input. Stencil e2e does not provide a way to trigger the blur event. Unable to find work around.

nathan-bird commented:
> Tests for update validation removed as validation does not happen until blur on input. Stencil e2e does not provide a way to trigger the blur event. Unable to find work around.

Since [Stencil 2.17.2](https://github.com/ionic-team/stencil/releases/tag/v2.17.2) they have added [a mock blur](https://github.com/ionic-team/stencil/pull/3449) to their event api. 
I suggest asking other Plumage devs on how best to try and use this api to write the tests you want.

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -153,60 +154,57 @@ export class Slider {
+      return this.updateValue(ev.target.value);
+    }
+    return this.updateValue(
+      this.allowedInputs.reduce((prev, curr) =>
```
No abbreviations 
```suggestion
      this.allowedInputs.reduce((previous, current) =>
```

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -153,60 +154,57 @@ export class Slider {
+    this.inputFieldValue = ev.target.value;
+  }
+
+  private validateInputField(ev) {
```
No abbreviations 
```suggestion
  private validateInputField(event) {
```
This ev abbreviation is seen throughout the component please fix all.

nathan-bird commented on packages/component-library/src/components/plmg-slider/test/plmg-slider.e2e.ts:
```diff
@@ -50,6 +50,15 @@ describe('plmg-slider sets the range and initial values', () => {
   });
 
   it('sets the value to the min if default value prop is above the allowed range', async () => {
```
I think something has gone wrong with your `main` merge since this test name does not relate to the test case.

nathan-bird reviewed (changes requested):
The input field is not populated with a value on render, it is only updated on slider change.
<img width="541" alt="Screenshot 2022-08-15 at 07 18 39" src="https://user-images.githubusercontent.com/14997231/184581185-91a9fd12-691c-4d43-993f-25f904606e41.png">

Keyboard navigation is  difficult/impossible.
- Hitting enter when focused on input field does not set the value
- Entering a number starts with an invalid value forces the user out of focus. To repeat; go to Stepped story, tab into the input field, try to type "15". After entering "1" focus will be removed and the slider will move to 5.

jamesducky commented:
> * Entering a number starts with an invalid value forces the user out of focus. To repeat; go to Stepped story, tab into the input field, try to type "15". After entering "1" focus will be removed and the slider will move to 5.

I think you'll only see this in storybook, not in the React example. The reason is the default keyboard shortcuts in storybook shift focus. 
![Screenshot 2022-08-15 at 08 27 59](https://user-images.githubusercontent.com/104433221/184587090-3355e4f5-a1f6-4587-9606-3f071bf533f1.png)

jamesducky commented:
incorrect commit message[90a032a](https://github.com/DuckyTeam/Plumage/pull/65/commits/90a032ae42da46edf772e6dff08dc7368fd36db0). This 
This fix sets falsy inputs in the field value to the slider's current value, not the min.

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -137,84 +154,106 @@ export class Slider {
    * The event "valueUpdated" is triggered when the slider value changes either by moving the thumb or entering in the text field.
    */
   @Event() valueUpdated: EventEmitter;
+  // /**
+  //  * The event "valueControlFailure" is emitted by when value passed
+  //  */
+  // @Event({ eventName: 'valueControlReject' })
+  // valueControlReject: EventEmitter;
```
Please delete this commented event handler.

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -137,84 +154,106 @@ export class Slider {
-  private handleSliderChange(ev) {
-    this.updateValue(ev.target.value);
+  private setAllowedInputs() {
+    const range = Array.from(
+      { length: (this.max - this.min) / this.stepValue + 1 },
+      (_, i) => this.min + i * this.stepValue
+    );
+    return range.map(
+      (value) => Math.round((value + Number.EPSILON) * 1000000) / 1000000
+    );
```
You have some fancy logic going on here. Could you explain what it is doing?

nathan-bird replied:
I would very much like to see your explanation 😃 I want to be sure that I (and you) completely understand what the code is doing and how.

jamesducky replied:
Decided to write a [blog post](https://dev.to/jamesducky/explained-range-generator-with-arrayfrom-3eh2-temp-slug-4862945?preview=c72a5cd08858c0b5aeeb6f3a4c0347e38c94481406cb53519aa318f1db7e02ad705f3372b27ab3632ecdf4c842ebd7509598a77b2639c291e18e1b4d) on the subject 😃

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -137,84 +154,106 @@ export class Slider {
+
   connectedCallback() {
-    this.setValues();
+    if (this.rangeValues) {
```
As said above, `rangeValues` should be required so this check isn't necessary.

nathan-bird reviewed (changes requested):
All in all a very good improvement to the component, the structure and logic is clear! 👏  
As we spoke about; the PR does encompass many different things and it would have been nice to see some of them as separate PRs but I understand the flow of making fixes and it didn't hinder my review.

I have some comments and suggestions for a few changes. Check em out. 

Great stuff 💪

jamesducky commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -125,6 +124,24 @@ export class Slider {
+   *
+   * Sets the value of the slider
+   */
+  @Prop() valueControl: number;
```
because I started introducing 'valueControl' so I tried to keep things consistent. Obviously I should have just renamed the prop to value.

jamesducky commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -83,13 +83,12 @@ export class Slider {
   @Prop() rangeValues: string;
```
agreed. now it will throw an error.

jamesducky commented on packages/component-library/src/components/plmg-slider/plmg-slider.tsx:
```diff
@@ -83,13 +83,12 @@ export class Slider {
-      throw new Error(
-        'rangeValues must be a comma separated list with at least two items'
-      );
+        this.stringToNumberArray(newValue).length < 2;
```
Changed. Now it will just throw an error if the array is
- not a string
- an empty string
- an array with less than two items
- if one of the items in the array is not a number

jamesducky commented on examples/react-app/src/Forms/SlidersForm.tsx:
```diff
@@ -0,0 +1,120 @@
+        display: 'grid',
+        gridTemplateColumns: '4fr 1fr',
+        width: '600px',
+        gap: '32px',
```
No good reason. I just thought it looked nice but it doesn't add much now I see it again. I've changed it back.

nathan-bird replied:
You're right, it did look nice and that is reason enough for me in an example app. However, I wanted to understand why you choose to use `grid` and use it in this way. 
Seeing the final design you were going for, I feel you could have styled it in a simpler way … that doesn't mean you have to, but you need to be able to explain why you choose the implementation you did.

jamesducky replied:
Oh, I see.  Basically I started with flex, didn't get the results I wanted and switched to grid. I can see now that I didn't need I think I just to tend to understand grid better than flex.

nathan-bird commented on examples/react-app/src/Forms/SlidersForm.tsx:
```diff
@@ -0,0 +1,143 @@
+    large: SLIDERS.large.default,
+  });
+
+  const handleChange = (value: number, target: HTMLElement) => {
+    const elementName = String(target.getAttribute('name'));
+    if (elementName) {
+      setSliders({ ...sliders, [elementName]: value });
+    }
+  };
+
+  const handleReset = (target: HTMLElement) => {
+    const elementLabel = target.getAttribute('aria-label');
+    if (elementLabel) {
+      setSliders({
+        ...sliders,
+        [elementLabel]: SLIDERS[elementLabel].default,
+      });
+    }
+  };
```
You have added two handler functions here which are calling the same thing; `setSlider`, which means they need the same information. I suggest simplifying and passing in your variable rather than cleverly deduce it from the element. How this form is implemented allows you to do that.
Remove one of the handers and pass in the slider identifier to be changed in state.  

Eg
```  
const handleChange = (value: number, slider: string) => {
    setSliders({ ...sliders, [slider]: value });
 };
 ```
Used via:
```
onValueUpdated={(event) =>
    handleChange(event.detail.value, SLIDERS.decimal.label)
}
onClick={() =>
    handleChange(SLIDERS.decimal.default, SLIDERS.decimal.label)
}
```

nathan-bird replied:
In fact, why create a handler at all and just call `setSliders` directly.

eg:
```
onValueUpdated={(event) => {
  setSliders({
    ...sliders,
    [SLIDERS.stepped.label]: event.detail.value,
  });
}}
onClick={() => {
  setSliders({
    ...sliders,
    [SLIDERS.stepped.label]: SLIDERS.stepped.default,
  });
}}
```

jamesducky replied:
Yes, this is obviously the simplest solution! I had thought it was bad practice to repeat the same function when it could be abstracted out, but abstracting it out just made things overly complicated.

I had thought about creating an array of components and then mapping over them but I decided that this would be less readable.

nathan-bird commented on common/changes/@ducky/plumage/fix-slider-2_2022-08-25-13-06.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@ducky/plumage",
```
Please ensure you add a patch changelog for `plumage-react` so that these fixes are released as part of that package.

jamesducky replied:
Think I've fixed that now

nathan-bird replied:
I'm slightly confused by what has been going with your change logs. 
Have you been manually changing the content or using what is generated by `rush change`?  
The final change logs have a type of `none`, is this what you intend? I would not expect so due to the commit message.

jamesducky replied:
I manually created the react change log file as it wouldn't generate. Clearly this was a stupid thing to do because I've introduced an error by doing this! I'll ask Tanguy about it and do it the right way.

jamesducky replied:
Files are now created automatically. I've set type to patch as I think this is minor update.

jamesducky commented:
Updates 
- merged react app example to include tab 
- changes tabs causes values to reset, so I've lifted state up to the parent component (tab) so now I'm passing the value update function as props to the child

nathan-bird commented on examples/react-app/src/FormElements/Sliders.tsx:
```diff
@@ -1,86 +1,124 @@
+};
 
+export default function SlidersForm(props: Props) {
+  const { sliders } = props;
+  const { onUpdate } = props;
```
You can consolidate your destructuring. 
```suggestion
  const { sliders, onUpdate } = props;
```

nathan-bird commented on examples/react-app/src/FormElements/Sliders.tsx:
```diff
@@ -1,86 +1,124 @@
+          default={SLIDERS.decimal.default}
+          marks={SLIDERS.decimal.marks}
+          step={SLIDERS.decimal.step}
+          range={SLIDERS.decimal.range}
+          value={sliders.decimal}
```
`SLIDERS` & `sliders` are the same identifier it can easily become confused. I would suggest a rename of one of them (or both) to be more explicit to what it is representing.

nathan-bird commented on examples/react-app/src/FormElements/Sliders.tsx:
```diff
@@ -1,86 +1,124 @@
+          Reset
+        </PlmgButton>
+        <label>{SLIDERS.stepped.label}</label>
+        <PlmgSlider
```
On first render I am seeing thumb label positioning problems.
<img width="1001" alt="Screenshot 2022-09-05 at 03 28 36" src="https://user-images.githubusercontent.com/14997231/188343118-313b38fb-6f98-4380-a9d0-77324622904b.png">
I have seen this happen on all the sliders, it does not appear to be unique to the decimal slider.

jamesducky replied:
![Screenshot 2022-09-05 at 09 04 40](https://user-images.githubusercontent.com/104433221/188389381-1cf21a19-56d9-4d24-b1c4-b755bc036939.png)
Managed to recreate the bug by putting a fixed width on the slider (240px).
The input field is actually set by the 'size' value, which is specified by the max value of the input number field. Forcing a min-width is making the overflow the container.
Using an outline is also causing overflow.
Fix: removed min-width, added a negative outline-offset on the input field 
![Screenshot 2022-09-05 at 09 07 20](https://user-images.githubusercontent.com/104433221/188393188-3c58334e-9cbb-4e95-9273-e6cf11b547fd.png)

jamesducky replied:
The 'track value' (the input range slider container's) width was getting changed between rendering the component container and calculating the thumb position because of flex gap and fit-content
(changing)[https://github.com/DuckyTeam/Plumage/pull/65/commits/66992c416685617a25b0ad7c94f24346e1a392e2] flex gap for margin-left with a fixed value and removing fit-content prevents input from rendering

nathan-bird commented on examples/react-app/src/FormElements/index.tsx:
```diff
@@ -1,10 +1,17 @@
+import { SLIDERS } from './Constants/Sliders';
 
 export default function FormElements() {
+  const [sliders, setSliders] = useState({
+    decimal: SLIDERS.decimal.default,
+    stepped: SLIDERS.stepped.default,
+    large: SLIDERS.large.default,
+  });
```
Why did you choose to move setting the slider state to the parent component?

jamesducky replied:
Because changing tabs was making the slider values reset. When the tabs changed, the SlidersForm component re-rendered causing the value to be reset. The alternative I could see was using a context provider.

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.stories.js:
```diff
@@ -54,42 +47,42 @@ Primary.args = {
   marks: true,
   name: 'range-slider',
-  ['default-value']: 5,
+  ['default']: 5,
```
```suggestion
  default: 5,
```

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.stories.js:
```diff
@@ -54,42 +47,42 @@ Primary.args = {
 export const ThumbLabel = Template.bind({});
 ThumbLabel.args = {
-  ['default-value']: 25,
+  ['default']: 25,
```
```suggestion
  default: 25,
```

nathan-bird commented on packages/component-library/src/components/plmg-slider/plmg-slider.stories.js:
```diff
@@ -54,42 +47,42 @@ Primary.args = {
 export const AllOptions = Template.bind({});
 AllOptions.args = {
-  ['default-value']: 0,
+  ['default']: 0,
```
```suggestion
  default: 0,
```

nathan-bird reviewed (commented):
A few comments for you to check out

jamesducky commented:
@nathan-bird: I've added a prop to allow the user to control textInputWidth. This is not because of the thumb positioning issue, which I've fixed separately, but I think we need this to control to maintain visual consistency.

nathan-bird reviewed (commented):
Updates look ok. Would be nice with a test of the new control.

nathan-bird reviewed (approved):
LGTM 🎉 

Next time something like this could be many smaller PRs which would make it faster to review and merge.
